File: Data Structures/Stack/Array implementation/java/Stack.java
import java.util.Scanner;

public class Stack {

    private int[] stack;
    private int size;
    private int top_pointer = -1;

    public Stack(int size) {
        this.size = size;
        this.stack = new int[this.size];
    }

    public boolean isEmpty() {
        return this.top_pointer == -1;
    }

    public boolean isFull() {
        return this.top_pointer == this.size-1;
    }

    public void push(int data) {
        if (!isFull())
            this.stack[++top_pointer] = data;
        else
            System.out.println("Stack overflow");
    }

    public int pop() {
        if (!isEmpty())
            return this.stack[this.top_pointer--];
        else {
            System.out.println("Stack underflow");
            return 0;
        }
    }

    public int top() {
        if (!isEmpty()) 
            return this.stack[this.top_pointer];
        else {
            System.out.println("Error: Stack is empty");
            return 0;
        }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.print("Enter the size of the stack: ");
        int size = sc.nextInt();

        Stack my_stack = new Stack(size);

        System.out.println("Enter 1 to push to stack");
        System.out.println("Enter 2 to pop from stack");
        System.out.println("Enter 3 to get top of stack");
        System.out.println("Enter 4 to check if the stack is empty");
        System.out.println("Enter 5 to check if the stack is full");
        System.out.println("Enter -1 to quit program");

        int option = 0;
        int push_data;

        while (option != -1) {
            System.out.print("Enter your option: ");
            option = sc.nextInt();

            switch (option) {
                case 1:
                    System.out.print("Enter data to push to stack: ");
                    push_data = sc.nextInt();
                    my_stack.push(push_data);
                    break;

                case 2:
                    System.out.println(my_stack.pop());
                    break;
                
                case 3:
                    System.out.println(my_stack.top());
                    break;

                case 4:
                    System.out.println(my_stack.isEmpty());
                    break;

                case 5:
                    System.out.println(my_stack.isFull());
                    break;

                case -1:
                    System.out.println("Program ended");
                    break;

                default:
                    System.out.println("Please enter a valid input");
            }
        }
        sc.close();
    }
}

File: Data Structures/Stack/Array implementation/python/stack.py
class Stack:
    def __init__(self, size):
        self.size = size
        self.stack = [None] * self.size
        self.top_pointer = -1

    def isEmpty(self):
        return self.top_pointer == -1

    def isFull(self):
        return self.top_pointer == self.size-1

    def push(self, data):
        if not self.isFull():
            self.top_pointer += 1
            self.stack[self.top_pointer] = data
        else:
            print("Stack overflow")

    def pop(self):
        if not self.isEmpty():
            temp = self.stack[self.top_pointer]
            self.top_pointer -= 1
            return temp
        else:
            print("Stack underflow")

    def top(self):
        if not self.isEmpty():
            return self.stack[self.top_pointer]
        else:
            print("Error: Stack is empty")

    
if __name__ == '__main__':
    size = int(input('Enter the size of stack: '))
    my_stack = Stack(size)

    print('Enter 1 to push to stack')
    print('Enter 2 to pop from stack')
    print('Enter 3 to get top of stack')
    print('Enter 4 to check if the stack is empty')
    print('Enter 5 to check if the stack is full')
    print('Enter -1 to quit program')

    option = 0
    push_data = None

    while option != -1:
        option = int(input('Enter your option: '))

        if option == 1:
            push_data = int(input("Enter data to push to stack: "))
            my_stack.push(push_data)

        elif option == 2:
                print(my_stack.pop())

        elif option == 3:
            print(my_stack.top())

        elif option == 4:
            print(my_stack.isEmpty())

        elif option == 5:
            print(my_stack.isFull())

        elif option == -1:
            print("Program ended")

        else:
            print("Please enter a valid input")
    